### ISSUE TITLE
my puzzle in the dataset of from height & weight to predict gender

### ISSUE THREAD

**Issue by charles-bu** [amueller/introduction_to_ml_with_python#76]
hi, 

I thought everyone would be eager to make some try after finishing reading your book. So do I.

I got one dataset of height & weight and gender, 3 columns and 10K samples for using the height & Weight to predict the gender.

before my going into the details, with your advice, I used your mglearn to plot the data distribution chart with X -axis for height and Y -axis for weight colored by different gender. 

I was a little shocked by the chart I made. I found both male and female data were almost in one block. therefore, I was in a little grey mood in continuing my ml learning code.

however, to my surprise, the result of the computing is quite promising. the svm.score on test set is 92.3% something,  which is totally out of my expectation. when I read your book, I got the concept of the dataset should have its own nature of being classified. the only thing for us is only to find a proper way to classify it, such as iris, or the breast cancer dataset, for instance,  in iris dataset, you used the pairplot to show us in certain features combination, we could see the data was in different group naturally.

however, this time, the whole dataset is almost in one single block. but the ML still could get a very good result, which made me totally lost. 

I will be grateful if you could provide me some advice. I really cannot pull me out of this puzzle.

meanwhile,as I did not find how to attach the dataset and my code, which is very small (less than 1M), may I have your advice of how to give it to you for your knowing my problem more clearly and accurately.

best regard,

Charles

**Comment by amueller** [comment]
Given that your data is only two-dimensional, you can in fact easily visualize it in it's totality with a scatter plot. The fact that it's in "one single block" could be the result of overplotting. Try to adjust alpha values or the marker size, or possibly look at hexbin plots to show the distribution more clearly.

You can just attach files to issues from what it looks like, you just need to zip them. Or you can post it anywhere else on the internet.

**Comment by charles-bu** [comment]
[Gender_v2.txt](https://github.com/amueller/introduction_to_ml_with_python/files/1863942/Gender_v2.txt)
[Uploading Gender_v2.txt…]()

**Comment by amueller** [comment]
that's the code, not the data, though...

**Comment by charles-bu** [comment]
here coming

**Comment by charles-bu** [comment]
[Gender.xlsx](https://github.com/amueller/introduction_to_ml_with_python/files/1863950/Gender.xlsx)

**Comment by amueller** [comment]
```python
import pandas as pd
from matplotlib.colors import BoundaryNorm
%matplotlib inline
import matplotlib.pyplot as plt
gender = pd.read_excel("Gender.xlsx")
gender.plot(x="Height", y="Weight", c="Gender", kind="scatter",
            alpha=.05, cmap=plt.cm.tab10, norm=BoundaryNorm([0, 1, 2], 2))
# don't worry about the cmap and norm, it's just to make the colors look decent
```
![image](https://user-images.githubusercontent.com/449558/38143254-a885b050-340d-11e8-86f7-2e6dc71a4813.png)
They are jointly approximately Gaussian, and each approximately Gaussian with similar variance but different means. That's about what I would expect.

**Comment by charles-bu** [comment]
hi, 

I have to convert the csv file into excel, as the system does not support  uploading the csv file. 

from the file, you may be able to see the most part of distribution of male and female overlapped. though both svm and mlp result  looks quite promising, I still can not persuade myself to accept the result. meanwhile, when I also used your function of mglearn.plots.plot_2d_separator,  it shows the classification border still in the middle of two data distribution. 

here I also attached the two of distribution chart of height and weight.
![height_distribution](https://user-images.githubusercontent.com/34530324/38143165-db462204-3471-11e8-9277-e0b69cdb904c.jpg)
![weight_distribution](https://user-images.githubusercontent.com/34530324/38143168-dc9408a6-3471-11e8-84ca-7d8b039b377a.jpg)



therefore, I got lost completely.

your advice is highly appreciated!


Charles

**Comment by amueller** [comment]
the problem with your plot / code is that you used train_test_split, then stacked the training and test set, and used the original labels. ``train_test_split`` shuffles the data so the labels in the plot were shuffled.

**Comment by amueller** [comment]
If you do
```python
y_plot = np.hstack((y_train, y_test))
mglearn.discrete_scatter(X_plot[:,0], X_plot[:,1], y_plot, 
                         alpha = 0.3)
```
that'll actually work.
But for plotting the scaling doesn't make a difference either way,
so you could also just have done
```python
mglearn.discrete_scatter(data_feature_np[:,0], data_feature_np[:,1], data_y_np, 
                         alpha = 0.3)
```

**Comment by charles-bu** [comment]
actually, I just tried code, 
mglearn.discrete_scatter(data_feature_np[:,0], data_feature_np[:,1], data_y_np, 
                         alpha = 0.3)
but the overlap is still has a large part.  sorry for the system continuous failed to upload my small picture.

**Comment by amueller** [comment]
As I mentioned earlier, that's an issue with overplotting. The distributions overlap but the main part where the mass is doesn't. Try playing with alpha.

**Comment by charles-bu** [comment]
the problem is done!

thank you so much for the help and Happy Easter!

Charles

**Comment by polytech1** [comment]
@charles-bu From where do these data come please ?
